src/pybind/mgr/dashboard/controllers/rbd.py

```diff
@@ -486,19 +512,47 @@ def purge(self, pool_name=None):
         for pool in pools:
             for image in pool['value']:
                 if image['deferment_end_time'] < now:
-                    _rbd_call(pool['pool_name'], self.rbd_inst.trash_remove, image['id'], 0)
+                    _rbd_call(pool['pool_name'], image['namespace'],
+                              self.rbd_inst.trash_remove, image['id'], 0)
 
-    @RbdTask('trash/restore', ['{pool_name}', '{image_id}', '{new_image_name}'], 2.0)
+    @RbdTask('trash/restore', ['{image_id_spec}', '{new_image_name}'], 2.0)
     @RESTController.Resource('POST')
     @CreatePermission
-    def restore(self, pool_name, image_id, new_image_name):
+    def restore(self, image_id_spec, new_image_name):
         """Restore an image from trash."""
-        return _rbd_call(pool_name, self.rbd_inst.trash_restore, image_id, new_image_name)
+        pool_name, namespace, image_id = parse_image_spec(image_id_spec)
+        return _rbd_call(pool_name, namespace, self.rbd_inst.trash_restore, image_id,
+                         new_image_name)
 
-    @RbdTask('trash/remove', ['{pool_name}', '{image_id}', '{image_name}'], 2.0)
-    def delete(self, pool_name, image_id, image_name, force=False):
+    @RbdTask('trash/remove', ['{image_id_spec}'], 2.0)
+    def delete(self, image_id_spec, force=False):
         """Delete an image from trash.
         If image deferment time has not expired you can not removed it unless use force.
         But an actively in-use by clones or has snapshots can not be removed.
         """
-        return _rbd_call(pool_name, self.rbd_inst.trash_remove, image_id, int(str_to_bool(force)))
+        pool_name, namespace, image_id = parse_image_spec(image_id_spec)
+        return _rbd_call(pool_name, namespace, self.rbd_inst.trash_remove, image_id,
+                         int(str_to_bool(force)))
+
+
+@ApiController('/block/namespace', Scope.RBD_IMAGE)
```

Comment:
Just to make sure I understood what you would like to change here...

This is the CRUD controller for RBD namespaces, equivalent to the following rbd commands:

- **rbd namespace create** _pool-name/namespace-name_
- **rbd namespace list** _pool-name_
- **rbd namespace remove** _pool-name/namespace-name_

See https://docs.ceph.com/docs/master/man/8/rbd/

Your sugestion it to change these endpoints, from:

- **POST** `/block/namespace` _body={pool_name:<p_name>, namespace:<n_name>}_ (create)
- **GET** `/block/namespace` (list)
- **DELETE** `/block/namespace/<pool_id>/<namespace_id>` (remove)

To:
- **POST** `/block/pool/<pool_id>/namespace` _body={namespace:<n_name>}_ (create)
- **GET** `/block/pool/<pool_id>/namespace` (list)
- **DELETE** `/block/pool/<pool_id>/namespace/<namespace_id>` (remove)

And change the UI to do the following requests:

- **GET** `/pool`
- For each pool
  - **GET** `/block/pool/<pool_id>/namespace`

Instead of doing 

- **GET** `/block/namespace`

to populate the following table:

![Screenshot from 2019-10-28 11-49-28](https://user-images.githubusercontent.com/14297426/67676562-b1bfb200-f979-11e9-842d-e844ca0939ee.png)

Is this what we want to change?

Reply:
Yes, that's what I was suggesting (rather than requesting, as I understand that it'd imply a great change in the way this has been done so far and would exceed the scope of this PR). 

I'm not a 100% decided on whether the pool aggregation should better happen in the Front-end or in the Back-end (each has pros & cons), but for delivering a REST API we should at least properly expose the resource hierarchy via the URL parts (`/block/pool/<pool_id>/namespace/<namespace_id>/image/<image_id>/snapshot/<snapshot_id>`).

Reply:
@epuertat I've now pushed one commit that improves the URL hierarchy as described in https://github.com/ceph/ceph/pull/30935#discussion_r339524564